Global Styles: Add block preview component in global styles

madhusudhand opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

This change adds block preview in global styles.

![allblocks-preview](https://user-images.githubusercontent.com/1935113/201289912-3bd5a4b9-f4da-46bf-8da8-dbe376598d3a.gif)

## How?

1. A new panel component has been introduced to include in the global styles.
2. The component considers the example from block-library and renders the component.
3. Supports enabled for few components to test the feature.

Note: there are few issues observed in the examples of block such as `image` which renders more spacing. However, it seem to be a different issue.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
1. Open site editor
4. Open global styles panel from the menu.
5. Go to Blocks -> [select any block such as paragraph]
6. View the preview of the block
7. Update the block styles such as typography, color etc..
8. View the new updates in the preview.

## Screenshots or screencast <!-- if applicable -->

![paragraph-preview](https://user-images.githubusercontent.com/1935113/201289926-02cf16ae-ec43-4d30-9305-d5f1009c76f3.gif)

## Issues

Partial fix for #42919

madhusudhand committed:
add block preview in global styles panel

update viewport widths for few blocks

add border to preview block and convert viewport width to scale

fix border radius and address minor review comments

address review comments

mtias commented on packages/edit-site/src/components/global-styles/hooks.js:
```diff
@@ -201,6 +201,11 @@ export function getSupportedGlobalStylesPanels( name ) {
 	}
 
+	// check for shadow support
+	if ( blockType?.supports?.preview ) {
```
Can we just assume this from the presence of an `example` construct?

madhusudhand replied:
Thanks for the review. Component preview is now based on the presence of example 👍

mtias commented:
Thanks, this is going to help a lot!

mikachan commented on packages/edit-site/src/components/global-styles/block-preview-panel.js:
```diff
@@ -0,0 +1,42 @@
+import { getBlockType, getBlockFromExample } from '@wordpress/blocks';
+import { useResizeObserver } from '@wordpress/compose';
+
+// const normalizedWidth = 248;
```
Should this comment be removed?

mikachan reviewed (commented):
This is looking great! It's testing well for me.

I've got a few comments:

- I'm not sure how obvious it is that this is a preview of the block. I think maybe a border around the block preview container would help make the preview boundary more obvious. (Maybe like the image on the right [here](https://user-images.githubusercontent.com/1204802/182621405-12247c5d-1e7f-45cc-87a5-57578b9b82f3.png)?)

- There seems to be a lot of padding on the iframe, especially at the top, e.g. the Quote block looks like it has a lot of space above its preview. I think making the preview area more obvious (e.g. with a border) may help this.
<img width="293" alt="image" src="https://user-images.githubusercontent.com/1645628/202728200-b66af55d-4542-4777-8c06-321fb43f1204.png">

- Some blocks don't preview anything, such as: List Item, Post Comments Link, Post Comments Count, Post Author Name, Post Title. I guess this is because there is no example content for them? (Although I don't think they all need to work for this to be merged, I was just highlighting it)

- The Navigation preview shows the mobile menu, but I'm unable to click on it to open it:
<img width="271" alt="image" src="https://user-images.githubusercontent.com/1645628/202728838-52f3c3c9-82e2-4aa4-9221-e8b43883ab3a.png">

- When there is no preview available, maybe we should show something like a disabled preview area, so it doesn't look like the preview is just broken/not showing? This could be a follow-up.

noisysocks commented on packages/block-editor/src/components/block-preview/auto.js:
```diff
@@ -68,7 +70,7 @@ function ScaledBlockPreview( {
 				maxHeight:
 					contentHeight > MAX_HEIGHT ? MAX_HEIGHT * scale : undefined,
-				minHeight: __experimentalMinHeight,
+				minHeight: `${ __experimentalMinHeight }px`,
```
This change is unnecessary, no? React will treat any number in the `style` object as a pixel value.

noisysocks commented on packages/edit-site/src/components/global-styles/block-preview-panel.js:
```diff
@@ -0,0 +1,42 @@
+
+	return ! blockExample ? null : (
+		<div
+			className="edit-site-block-styles-preview__panel"
```
This class name isn't following BEM properly. It should be `edit-site-global-styles__block-preview-panel`.

noisysocks commented on packages/edit-site/src/components/global-styles/block-preview-panel.js:
```diff
@@ -0,0 +1,42 @@
+	return ! blockExample ? null : (
+		<div
+			className="edit-site-block-styles-preview__panel"
+			style={ panelStyles }
```
None of this styles seem to change at runtime. Can we move them to `style.scss`?

```scss
.edit-site-global-styles__block-preview-panel {
	width: 100%;
	height: 152px;
	overflow: auto;
}
```

madhusudhand replied:
Moved all the styles to class, except for height because the variable is used twice in JS.

madhusudhand replied:
Moved the height also to a `scss` variable instead.

noisysocks commented on packages/edit-site/src/components/global-styles/block-preview-panel.js:
```diff
@@ -0,0 +1,42 @@
+			className="edit-site-block-styles-preview__panel"
+			style={ panelStyles }
+		>
+			<div style={ { position: 'relative', width: '100%', height: 0 } }>
```
What is this container for? It's odd to have a div with 0 height. Could `containerResizeListener` live directly in `.edit-site-block-styles-preview__panel`?

madhusudhand replied:
removed the container 👍

noisysocks commented on packages/block-library/src/audio/index.js:
```diff
@@ -23,6 +23,7 @@ export const settings = {
 		attributes: {
 			src: 'https://upload.wikimedia.org/wikipedia/commons/d/dd/Armstrong_Small_Step.ogg',
 		},
+		viewportWidth: 350,
```
Could you explain a little why all of these changes to the `example` in `packages/block-library` are necessary? Not sure I understand.

madhusudhand replied:
The `BlockPreview` component scales the content with given viewportWidth to fit large blocks relative to given container.
I agree, these numbers are confusing. I introduced much simpler scaling approach (ex: 0.75) [here](https://github.com/WordPress/gutenberg/pull/45719/files#diff-ba97676c79b4678b29dcdb77d3812f7b160b824df934c1289994831df2ce0c0fR65) which is much simpler.

This is required for blocks such as `audio`, to fit within the container.

noisysocks replied:
Aha, I see. Is this a new API?

madhusudhand replied:
It's an existing API. Here is the ref from [the docs](https://github.com/WordPress/gutenberg/pull/45719/files#diff-a2de963e686384ef0f566d26844f206f08969e704113bd4a9f45151ff629c567R219)

With the introduction of `scale`, it is no longer used. Marking the conversation resolved.

noisysocks reviewed (commented):
Thanks for working on this! I focused on reviewing the code since @mikachan already has left great feedback on the functionality.

madhusudhand commented:
@mikachan @noisysocks Thanks for the review.

> I think maybe a border around the block preview container would help make the preview boundary more obvious.

Introduced the boundary to make it obvious.
<img width="274" alt="image" src="https://user-images.githubusercontent.com/1935113/203023722-0d694571-0cd7-4720-81ca-e63f087e0bbf.png">

> There seems to be a lot of padding on the iframe, especially at the top

This is due to double spacing ( container padding and block vertical margin ). Addressed it using defaultPadding from example which is set to only a few blocks.

> Some blocks don't preview anything, such as: List Item

Right, because they doesn't have an example. They can be added in a separate PR.

> The Navigation preview shows the mobile menu, but I'm unable to click on it to open it

The preview is not interactive. It doesn't show hover state for button for example.
However, Example could be revised to show desktop nav items in the preview. I would like to consider this when other examples are added.

> When there is no preview available, maybe we should show something like a disabled preview area, so it doesn't look like the preview is just broken/not showing? This could be a follow-up.

mmm, makes sense and it will be consistent. It could be a follow-up

jasmussen commented:
> Introduced the boundary to make it obvious.

Can you add a `border-radius: $radius-block-ui;` so it gets 2px rounded corners?

mtias commented:
I think we should also use some opacity or blend mode on the border so that it works ok in situations with custom background colors.

noisysocks commented on packages/block-editor/src/components/block-preview/auto.js:
```diff
@@ -59,7 +62,14 @@ function ScaledBlockPreview( {
 	const scale = containerWidth / viewportWidth;
+
+	// support number and string padding
+	const padding = isNaN( __experimentalPadding )
```
Minor but I'd suggest `typeof __experimentalPadding === 'string'` reads nicer than `isNaN` (you could even remove the comment) and will cause less grief if this code is ever converted to TypeScript (because `isNaN` has type `( x : number ) => boolean`).

noisysocks commented on packages/edit-site/src/components/global-styles/style.scss:
```diff
@@ -117,3 +117,10 @@
+	width: 100%;
+	overflow: auto;
+	border: $gray-200 $border-width solid;
+	box-sizing: content-box !important;
```
Is the `!important` necessary? It's best to avoid `!important` if possible as it's difficult for plugins to override.

madhusudhand replied:
Removed the style definition. Created a `scss` variable to manage the height instead.

noisysocks commented on packages/edit-site/src/components/global-styles/block-preview-panel.js:
```diff
@@ -0,0 +1,40 @@
+		<div
+			className="edit-site-global-styles__block-preview-panel"
+			style={ {
+				position: 'relative',
```
Should this go into `style.scss` as well?

noisysocks reviewed (commented):
Does setting `scale` and `defaultPadding` have any effect for the inserter's block preview? or is it just for this new instance of the block preview?

Should we update the block API documentation to document `scale` and `defaultPadding`?

madhusudhand commented:
@noisysocks 

> Does setting scale and defaultPadding have any effect for the inserter's block preview? or is it just for this new instance of the block preview?

No. They are optional. Inserter block preview doesn't define them and so doesn't have any effect. Only preview in styles panel use them.

> Should we update the block API documentation to document scale and defaultPadding?

Updated the docs with examples [here](https://github.com/WordPress/gutenberg/pull/45719/files#diff-a2de963e686384ef0f566d26844f206f08969e704113bd4a9f45151ff629c567R182)

madhusudhand commented:
@mtias 

> I think we should also use some opacity or blend mode on the border so that it works ok in situations with custom background colors.

I have done some testing with different colors with the current changes and following are the screenshots. 

<img width="281" alt="image" src="https://user-images.githubusercontent.com/1935113/203285736-f11a5c55-8ff1-43d6-977d-2df994192a77.png"> <img width="281" alt="image" src="https://user-images.githubusercontent.com/1935113/203285661-1d005f1e-f492-451b-8df5-eb06cd46bb14.png">
<img width="281" alt="image" src="https://user-images.githubusercontent.com/1935113/203285542-3edcea20-9b21-4bfd-b795-f9c1f06efa30.png"> <img width="281" alt="image" src="https://user-images.githubusercontent.com/1935113/203285303-69420864-c4b8-4f84-b866-8fc7615847c3.png">

When a dark background is set, the border is almost not visible. 
It plays a role to highlight the preview container, when a lighter background is set. 
It is the same as that of the style variations preview.

I might want to try the blend-mode for the border on both of these together in a separate PR.

noisysocks commented on docs/reference-guides/block-api/block-registration.md:
````diff
@@ -216,7 +216,7 @@ example: {
 ```
 
-It's also possible to define the width of the preview container in pixels via `viewportWidth`. For example:
+It's also possible to define the width of the preview container in pixels via `viewportWidth` or can specify `scale` to set width of the block relative to the preview container. For example:
````
```suggestion
It's also possible to define the width of the preview container in pixels via `viewportWidth` or via `scale` to set width of the block relative to the preview container. For example:
```

noisysocks commented on packages/base-styles/_variables.scss:
```diff
@@ -127,3 +127,6 @@ $block-selected-border-width: 1px;
 $block-selected-to-content: $block-edge-to-content - $block-selected-margin - $block-selected-border-width;
+
+// Block Preview
+$block-preview-height: 150px;
```
I like the variable as it's self-documenting but I don't imagine that it will be used much elsewhere. I'd move it to `edit-site/src/components/global-styles/style.scss`.

ntsekouras commented on packages/block-editor/src/components/block-preview/auto.js:
```diff
@@ -83,12 +93,23 @@ function ScaledBlockPreview( {
 					bodyElement.style.width = '100%';
+					bodyElement.style.minHeight = `${ __experimentalMinHeight }px`;
+
+					if ( __experimentalAlign || __experimentalJustify ) {
```
Why do we need these styles? What are the use cases for `display:flex`?

madhusudhand replied:
it is to align the block with in the preview container, both horizontally and vertically. Its is done conditionally no not effect `Block Inserter` preview.

<img width="181" alt="image" src="https://user-images.githubusercontent.com/1935113/203524517-dda88585-cdc1-444d-a41e-39c8c247b1d5.png">

ntsekouras replied:
Can we do it with css only like we do for example in block patterns preview in patterns explorer? I'm not 100% sure it's the same thing we want, but it seems like it: 
<img width="1785" alt="Screenshot 2022-11-23 at 12 51 30 PM" src="https://user-images.githubusercontent.com/16275880/203528815-be49d6d4-d979-4ab3-926e-857499936c54.png">

madhusudhand replied:
Container is not set 100% height here and alignment is taken care in parent. But in preview, full height is required and so flex alignment within preview panel.

<img width="1080" alt="image" src="https://user-images.githubusercontent.com/1935113/203974296-6c10d7d9-4f81-4d09-849d-5aa80ad7ce32.png">

Reverted the change from this PR and will take it to second iteration.

ntsekouras commented on packages/block-editor/src/components/block-preview/auto.js:
```diff
@@ -23,8 +23,11 @@ const MAX_HEIGHT = 2000;
+	__experimentalScale,
 	__experimentalPadding,
 	__experimentalMinHeight,
+	__experimentalAlign,
```
I'd be a bit cautious about adding so many new props here. It can easily get out of hand if we add props for every css prop we might need. Should we add something like `bodyStyles` and leave the logic to the consumers?

madhusudhand replied:
mmm, right. It's better to introduce a general property. 👍

ntsekouras commented on packages/block-library/src/table/index.js:
```diff
@@ -91,6 +91,8 @@ export const settings = {
 			],
 		},
+		defaultPadding: '1rem',
+		scale: 0.65,
```
How do you calculate these values of the new props for each block? Why do we need them in some blocks only?

madhusudhand replied:
baseline is `1` (actual size), and `0.65` means shrink to 65% of its original size.
Only few blocks such as `table`, `search` require this because their minimus size is way bigger than preview area.

ntsekouras replied:
What about `defaultPadding`?

madhusudhand replied:
it was to handle the space around the block. Reverted the api and set a default margin of 1rem in the CSS instead.

ntsekouras commented on docs/reference-guides/block-api/block-registration.md:
```diff
@@ -227,6 +227,26 @@ example: {
+	attributes: {
+		cover: 'https://example.com/image.jpg',
+	},
+	scale: 0.8
```
Can't we achieve the same result with the existing `viewportWidth` somehow?

madhusudhand replied:
It can be achieved with `viewportWidth`, but numbers are confusing. Preview panel is `150px`, however to fit the block properly, it requires numbers such as `400` or `500` which is very confusing.
The `scale` is much simpler which defaults to `1` meaning actual size and `0.8` means shrink to `80%` of the actual size.

ntsekouras replied:
>it requires numbers such as 400 or 500 which is very confusing.

I think that's subjective and `viewportWidth` is used for a while now. 🤷 It seems we try to accommodate for the specific height of this preview that overflows and the others don't?

noisysocks replied:
Yeah I've mixed feelings on this. On the one hand `scale` is arguably a nicer API, but on the other hand you can already accomplish this with `viewportWidth` which exists already. We have to be really careful when adding new APIs to `block.json` because once they're in Core we basically have to maintain them forever. (Core has a very strict backwards compatibility policy.)

mtias replied:
Yeah, let's try to leverage what exists. If we need to allow for scale outside of viewportWidth the name should also be viewportScale.

madhusudhand replied:
Thanks for the wonderful feedback. Reverted the new APIs and making use of `viewportWidth`

ntsekouras commented on packages/block-editor/src/components/block-preview/auto.js:
```diff
@@ -59,7 +62,14 @@ function ScaledBlockPreview( {
 	// Initialize on render instead of module top level, to avoid circular dependency issues.
 	MemoizedBlockList = MemoizedBlockList || pure( BlockList );
 
+	viewportWidth = viewportWidth / ( __experimentalScale || 1 );
```
This is quite confusing to me. We already calculate a `scale` (right below) and we still pass a new `scale` that takes part of the calculation. What was the need for this?

madhusudhand replied:
scale prop `__experimentalScale` is an alternative to reduce `viewportWidth` by certain percentage. line below takes scaled viewportWidth for the actual calculation and used everywhere. 
Perhaps the line below needs to be renamed?

madhusudhand replied:
change reverted.

ntsekouras reviewed (commented):
Thanks so much for your work @madhusudhand ! I have left some questions to better understand the approach here 😄

noisysocks commented:
Functionality wise, this is working really well. Nice job!

https://user-images.githubusercontent.com/612155/203667994-9b00265c-fcce-47ed-ac18-a81027369c82.mp4

noisysocks commented:
P.S. I added `Needs Dev Note` as there are new `block.json` APIs. If they're removed then we don't need a dev note.

madhusudhand committed:
revert scale

madhusudhand committed:
revert defaultPadding and alignment APIs

madhusudhand commented:
> P.S. I added Needs Dev Note as there are new block.json APIs. If they're removed then we don't need a dev note.

Reverted the new APIs. This PRs has [this tiny change](https://github.com/WordPress/gutenberg/pull/45719/files#diff-a2de963e686384ef0f566d26844f206f08969e704113bd4a9f45151ff629c567R182) in docs.
Not sure if this is still relevant.

Thanks for the wonderful feedback. Suggestions are addressed.

Note: Took a change related to center alignment of the block out of this PR as it needs to be commonly addressed in all the preview components such as `Block Inserter`. Will address it in a separate PR.

ntsekouras commented on packages/block-editor/src/components/block-preview/style.scss:
```diff
@@ -35,9 +35,15 @@
+		display: none;
+	}
+
+	.block-editor-block-list__layout {
```
It's better to add this style change only to the panel we want it now. If we need to update the styles for all BlockPreviews we should do it in a separate PR.

madhusudhand replied:
Agree that, It would be easier test all Previews in a separate PR.

However, this style is to have margin inside of iframe (created by BlockPreview) and cannot be targeted from outside. And JS prop just for margin isn't a good idea. So, made this CSS part of this PR. Suggest if we should ship this change without a default spacing (preview below)

<img width="269" alt="image" src="https://user-images.githubusercontent.com/1935113/204252882-d5b76ff4-969b-4936-b589-fdf84bff9dc3.png">

ntsekouras replied:
I think that's definitely something to be skipped from this PR. Also the `block-editor-block-list__layout` class is being added not only in the root container but also in every block that has inner blocks like `list, group` etc.. adding extra margins. This could be solved with a different selector, but still it affects all previews.

Personally I think without the margin this still is valuable addition.

madhusudhand replied:
Removed the change from this PR. 👍

ntsekouras commented on packages/block-library/src/paragraph/index.js:
```diff
@@ -30,12 +30,7 @@ export const settings = {
-				},
-			},
-			dropCap: true,
+			dropCap: false,
```
Nit: If we want to remove `dropCap` we can just omit(default is false).

ntsekouras reviewed (commented):
Thanks for iterating on this @madhusudhand! This looks close!

I'm not sure if this was discussed before, but do we really want the extra margin and in general so much spacing in the previews? If we for example add background it looks like this:
<img width="289" alt="Screenshot 2022-11-25 at 4 03 40 PM" src="https://user-images.githubusercontent.com/16275880/204002344-cc815387-cb87-4196-b530-40ba372adea7.png">

I'll leave this for the designers 😄 --cc @jameskoster

jameskoster commented:
> I'm not sure if this was discussed before, but do we really want the extra margin and in general so much spacing in the previews? If we for example add background it looks like this:

Good point, the margin can make the display a bit awkward, especially if you have something like a Quote where the margin is so large that it pushes the text out of the preview viewport. Does the preview need to be full-size, or can we scale it like the previews in the Inserter?

A side note: Does the preview window background inherit from global styles, or is it hard-coded to be white?

madhusudhand commented:
> I'm not sure if this was discussed before, but do we really want the extra margin and in general so much spacing in the previews? If we for example add background it looks like this:

The extra space is not entirely due to the margin. It is margin + padding set by custom CSS in specific themes, to add root level padding. Now that #35607 is fixed, individual themes needs a fix to remove the extra padding.

With no such extra padding in themes such as Twenty-Twenty Three, the preview with background is similar to how it looks in the editor, and it is expected I guess. (theme: TT3)

<img width="686" alt="image" src="https://user-images.githubusercontent.com/1935113/204258773-1c26609a-316e-4ee2-9e5a-3b006df12524.png">

> A side note: Does the preview window background inherit from global styles, or is it hard-coded to be white?

It inherits from the global styles.

<img width="686" alt="image" src="https://user-images.githubusercontent.com/1935113/204261327-936170f0-0bc2-4f04-8084-504fb1b70f26.png">

> Does the preview need to be full-size, or can we scale it like the previews in the Inserter?

With scaling it is no longer `as-is` view for typography specially for `font-size`. Hence it is best to keep the original scale.
However, few blocks such as `Table` are scaled down, as they do not fit within the preview.

madhusudhand committed:
remove dropcap for paragraph

jasmussen commented:
Took this for a quick spin, looks good, thanks for pushing this one. Even if there are rough edges, it brings so many benefits to it so it would be good to land this even if we need to follow up with some iterations.

Two quick observations. It might be good to set box-sizing: border-box; on the block when inside the preview, so it doesn't break out of boundaries when applying border:

<img width="305" alt="Screenshot 2022-12-06 at 13 38 00" src="https://user-images.githubusercontent.com/1204802/205919245-f0fb0ad3-a049-4130-a610-a79df838df0f.png">

And secondly, margin really doesn't translate well in the small preview:

<img width="351" alt="Screenshot 2022-12-06 at 14 00 15" src="https://user-images.githubusercontent.com/1204802/205919484-42ce2bf1-92f7-404f-8dc3-d3bc10f0c5b0.png">

It might be useful to zero out all margins in this preview. It isn't big enough to show the effect anyway. If need be we could explore a followup showing a larger preview on hover, something like that.

ntsekouras commented:
>It might be useful to zero out all margins in this preview. It isn't big enough to show the effect anyway. If need be we could explore a followup showing a larger preview on hover, something like that.

I think that's a complex issue to solve holistically and could lead to unexpected results. For example even if we add the params/styles to zero out margins and/or padding, there could be blocks that theme styles add borders etc..(ex. `code` in 2022).

jasmussen commented:
Nothing to block this PR, as noted, it's such a cool feature and it would be nice to see it soon!

ntsekouras reviewed (approved):
Thanks for iterating on this one @madhusudhand! Code wise looks good and IMO we can land and iterate to improve. If there is no objection from designers to land this as first version let's 🚢

mtias commented:
Nice work @madhusudhand !

bph commented:
Note for my future self: double-checked on the Dev note label. [Block api changes](https://github.com/WordPress/gutenberg/pull/45719#issuecomment-1323401423) for `scale` and `defaultPadding` were removed as per [later comment ](https://github.com/WordPress/gutenberg/pull/45719#issuecomment-1327362630)
